components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfSignUpHandler.java

```diff
@@ -0,0 +1,232 @@
+import org.wso2.carbon.identity.recovery.RecoverySteps;
+import org.wso2.carbon.identity.recovery.mapping.SelfSignUpConfig;
+import org.wso2.carbon.identity.recovery.model.Property;
+import org.wso2.carbon.identity.recovery.model.UserRecoveryData;
+import org.wso2.carbon.identity.recovery.store.JDBCRecoveryDataStore;
+import org.wso2.carbon.identity.recovery.store.UserRecoveryDataStore;
+import org.wso2.carbon.identity.recovery.util.Utils;
+import org.wso2.carbon.lcm.core.LifecycleOperationManager;
+import org.wso2.carbon.lcm.core.exception.LifecycleException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.mgt.constant.StoreConstants.IdentityStoreConstants;
+import static org.wso2.carbon.identity.mgt.constant.StoreConstants.IdentityStoreInterceptorConstants;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.CONFIRMATION_CODE;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.ErrorCodes;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.ErrorCodes.ACCOUNT_UNVERIFIED;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.NOTIFICATION_TYPE_ACCOUNT_CONFIRM;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.SELF_SIGN_UP_EVENT;
+import static org.wso2.carbon.identity.recovery.IdentityRecoveryConstants.SELF_SIGN_UP_PROPERTIES;
+
+/**
+ * User SelfRegistration Handler.
+ */
+public class UserSelfSignUpHandler extends AbstractEventHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(UserSelfSignUpHandler.class);
+
+    private SelfSignUpConfig config;
+    private RealmService realmService;
+    private EventService eventService;
+
+    @Override
+    public void configure(InitConfig initConfig) throws IdentityException {
+    }
+
+    public String getName() {
+        return "user.self.sign.up";
+    }
+
+    @Override
+    public void handle(EventContext eventContext, Event event) throws IdentityException {
+
+        // TODO: Use the new config model.
+        config = new SelfSignUpConfig();
+        config.setNotificationInternallyManaged(false);
+        config.setSelfSignUpEnabled(true);
+        config.setAccountLockOnCreation(true);
+
+        if (SELF_SIGN_UP_EVENT.equals(event.getEventName())) {
+
+            if (!config.isSelfSignUpEnabled()) {
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Self sign-up is disabled in the configuration.");
+                }
+                return;
+            }
+
+            Map<String, Object> eventProperties = event.getEventProperties();
+            User user = (User) eventProperties.get(IdentityStoreConstants.USER);
+
+            if (user == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Event property: {} cannot be found in event: {}.", IdentityStoreConstants.USER,
```

Comment:
Looks like "event: {}." is filled with ErrorCode. either, 1. Change the log message, 2. Add Event details or 3. do 1 and 2 both.